src/sumo_store_riak.erl

```diff
@@ -0,0 +1,421 @@
+
+%% Public API.
+-export([init/1]).
+-export([create_schema/2]).
+-export([persist/2]).
+-export([delete_by/3, delete_all/2]).
+-export([find_all/2, find_all/5, find_by/3, find_by/5, find_by/6]).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Types.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-record(state, {conn :: pid(), bucket :: binary(), index :: binary()}).
+-type state() :: #state{}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% External API.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-spec init(
+  term()
+) -> {ok, term()}.
+init(Options) ->
+  % The storage backend key in the options specifies the name of the process
+  % which creates and initializes the storage backend.
+  Backend = proplists:get_value(storage_backend, Options),
+  State = sumo_backend_riak:get_state(Backend),
+  {ok, State}.
+
+-spec persist(
+  sumo_internal:doc(), state()
+) -> sumo_store:result(sumo_internal:doc(), state()).
+persist(Doc, #state{conn = Conn, bucket = Bucket} = State) ->
+  {Id, NewDoc} = new_doc(Doc, State),
+  case update_map(Conn, Bucket, Id, doc_to_rmap(NewDoc)) of
+    {error, Error} ->
+      {error, Error, State};
+    _ ->
+      {ok, NewDoc, State}
+  end.
+
+-spec delete_by(
+  sumo:schema_name(), sumo:conditions(), state()
+) -> sumo_store:result(sumo_store:affected_rows(), state()).
+delete_by(DocName,
+          Conditions,
+          #state{conn = Conn, bucket = Bucket, index = Index} = State) ->
+  IdField = sumo_internal:id_field_name(DocName),
+  case lists:keyfind(IdField, 1, Conditions) of
+    {_K, Key} ->
+      case delete_map(Conn, Bucket, iolist_to_binary(Key)) of
+        ok ->
+          {ok, 1, State};
+        {error, Error} ->
+          {error, Error, State}
+      end;
+    _ ->
+      Query = build_query(Conditions),
+      case search_docs_by(DocName, Conn, Index, Query, 0, 0) of
+        {ok, {Total, Res}}  ->
+          delete_docs(Conn, Bucket, Res),
+          {ok, Total, State};
+        {error, Error} ->
+          {error, Error, State}
+      end
+  end.
+
+-spec delete_all(
+  sumo:schema_name(), state()
+) -> sumo_store:result(sumo_store:affected_rows(), state()).
+delete_all(_DocName, #state{conn = Conn, bucket = Bucket} = State) ->
+  Del = fun({C, B, Kst}, Acc) ->
+          lists:foreach(fun(K) -> delete_map(C, B, K) end, Kst),
+          Acc + length(Kst)
+        end,
+  case stream_keys(Conn, Bucket, Del, 0) of
+    {ok, Count} -> {ok, Count, State};
+    {_, Count}  -> {error, Count, State}
+  end.
+
+-spec find_all(
+  sumo:schema_name(), state()
+) -> sumo_store:result([sumo_internal:doc()], state()).
+find_all(DocName, #state{conn = Conn, bucket = Bucket} = State) ->
+  Get = fun({C, B, Kst}, Acc) ->
+          Fun = fun(K, Acc0) ->
+                  case fetch_map(C, B, K) of
```

Comment:
The expression on line 128 and column 19 is nested beyond the maximum level of 3.